src/tokens/tokens.service.ts

```diff
@@ -0,0 +1,217 @@
+    this.baseUrl = baseUrl;
+    this.instanceUrl = instanceUrl;
+    this.identity = identity;
+  }
+
+  private get postOptions() {
+    return {
+      params: {
+        'fly-from': this.identity,
+        'fly-sync': 'false',
+      },
+    };
+  }
+
+  async getReceipt(id: string): Promise<EventStreamReply> {
+    const response = await this.http
+      .get<EventStreamReply>(`${this.baseUrl}/reply/${id}`)
+      .toPromise();
+    return response.data;
+  }
+
+  async createPool(dto: TokenPool): Promise<AsyncResponse> {
+    const response = await this.http
+      .post<EthConnectAsyncResponse>(
+        `${this.instanceUrl}/create`,
+        {
+          uri: packTokenUri(dto.namespace, dto.name, dto.client_id),
+          is_fungible: dto.type === TokenType.FUNGIBLE,
+        },
+        this.postOptions,
+      )
+      .toPromise();
+    return { id: response.data.id };
+  }
+
+  async mint(dto: TokenMint): Promise<AsyncResponse> {
+    const type_id = packTokenId(dto.pool_id);
+    if (isFungible(dto.pool_id)) {
+      const response = await this.http
+        .post<EthConnectAsyncResponse>(
+          `${this.instanceUrl}/mintFungible`,
+          {
+            type_id,
+            to: [dto.to],
+            amounts: [dto.amount],
+            data: [0],
+          },
+          this.postOptions,
+        )
+        .toPromise();
+      return { id: response.data.id };
+    } else {
+      const to: string[] = [];
+      for (let i = 0; i < dto.amount; i++) {
+        to.push(dto.to);
+      }
+
+      const response = await this.http
+        .post<EthConnectAsyncResponse>(
+          `${this.instanceUrl}/mintNonFungible`,
+          {
+            type_id,
+            to,
+            data: [0],
+          },
+          this.postOptions,
+        )
+        .toPromise();
+      return { id: response.data.id };
+    }
+  }
+
+  async balance(dto: TokenBalanceQuery): Promise<TokenBalance> {
+    const response = await this.http
+      .get<EthConnectReturn>(`${this.instanceUrl}/balanceOf`, {
+        params: {
+          account: dto.account,
+          id: packTokenId(dto.pool_id, dto.token_index),
+        },
+      })
+      .toPromise();
+    return { balance: parseInt(response.data.output) };
+  }
+
+  async transfer(dto: TokenTransfer): Promise<AsyncResponse> {
+    const response = await this.http
+      .post<EthConnectAsyncResponse>(
+        `${this.instanceUrl}/safeTransferFrom`,
+        {
+          from: dto.from,
+          to: dto.to,
+          id: packTokenId(dto.pool_id, dto.token_index),
+          amount: dto.amount,
+          data: [0],
+        },
+        this.postOptions,
+      )
+      .toPromise();
+    return { id: response.data.id };
+  }
+}
+
+class TokenListener implements EventListener {
+  private readonly logger = new Logger(TokenListener.name);
+
+  constructor(private proxy: EventStreamProxyGateway) {}
+
+  handleEvent(event: Event) {
```

Comment:
I was looking for what ensures we process these sequentially, and I'm not seeing it here.
Is it that there's actually no `async` processing here? So everything will be ordered on the event loop?

E.g. the processing looks like the following, and none of it is async so we don't need any sequencing:

```
- Array[28] events arrives from ethconnect
- Repeat 28 times:
  - Synchronously invoke `handleEvent` (via listener registered)
  - `handleEvent` calls `broadcast` - guaranteed to be the next event to FireFly because of the semantics of `broadcast`
  - Mini `ack()` goes to common handler
- `Ack` goes back to ethconnect, because we're at 28 now
- Only now should we receive the next package
```

One edge case I know is possible, is on disconnect/reconnect to Ethconnect (if the websocket cycles) you might be re-driven with the same package of events again. That would be parallel.

Reply:
Correct, I'm just relying on synchronous code to guarantee order of delivery. Had not thought about the reconnect case while the previous event loop is still in flight...

All of this again calls into question what's the best way to handle batching. Again the current behavior of receiving the messages in batches but unpacking them during translation is beginning to feel wrong.

Reply:
Changed this up in the latest. Added some handling for client (ie FireFly) reconnects... not totally sure if I still need some additional handling for server (ie event stream) reconnects.

Reply:
Thinking through this again, we'll never get a reconnect interleaved with processing an event batch, because the batch processing loop is synchronous (it will always run to completion). So ordering will always be consistent. In a server-side reconnect we may receive (and proxy) some duplicate events, but the new ack behavior should ensure acks are matched to the correct events and duplicate acks are ignored.